### PR TITLE
AP_Airspeed: remove positive bias near zero differential pressure

### PR DESCRIPTION
**Pull request by kd0aij**
move sign flip and positive clamping after lowpass IIR filter.
Improves airspeed accuracy when average differential pressure is less than the noise amplitude.

### CODE REVIEW

**Comment by tridge** [comment]
@kd0aij to get the AIRSPEED_AUTOCAL message you need to set the parameter ARSPD_AUTOCAL=1

If you want to send it even when not flying (for bench testing) you need a change like this one:
https://github.com/tridge/ardupilot/commit/16bf874248a1e4b272eeeeb7110953a6c518209c

**Comment by tridge** [comment]
get_differential_pressure() can now return a -ve number, which means the call in Airspeed_Autocal.cpp needs a fabsf().
That is the only problem I've found, apart from that this looks good, thanks!

**Comment by kd0aij** [comment]
Initial calibration of 5525 sensor resulted in: ARSPD_OFFSET     53.843128
After 8 hours running on the bench, airspeed is reported as approx. 1 m/sec:
8hours_aftercal_5525

Rerunning airspeed cal then resulted in: ARSPD_OFFSET     54.634380
so the offset drift for this sensor was about 0.8 over 8 hours. Also note that the calibration routine is functioning correctly: the average (differential pressure - offset) is close to zero after recalibrating.

recalibrate_5525

File: ArduPlane/ArduPlane.cpp [hunk 1]
```diff
@@ -435,7 +438,7 @@ void Plane::airspeed_ratio_update(void)
         // don't calibrate when going beyond normal flight envelope
         return;
     }
-    const Vector3f &vg = gps.velocity();
+//    const Vector3f &vg = gps.velocity();
```
Comment (magicrub) [inline ArduPlane/ArduPlane.cpp]: please remove instead of commenting out

File: ArduPlane/ArduPlane.cpp [hunk 2]
```diff
@@ -67,7 +67,7 @@ const AP_Scheduler::Task Plane::scheduler_tasks[] = {
     SCHED_TASK(check_long_failsafe,     3,    400),
     SCHED_TASK(read_receiver_rssi,     10,    100),
     SCHED_TASK(rpm_update,             10,    100),
-    SCHED_TASK(airspeed_ratio_update,   1,    100),
```
Comment (magicrub) [inline ArduPlane/ArduPlane.cpp]: what's the reason for switching to 10Hz?

**Comment by magicrub** [comment]
interesting test. How does that compare to before the change?

File: ArduPlane/ArduPlane.cpp (hunk 1 above)
Comment (tridge) [reply ArduPlane/ArduPlane.cpp]: that patch shouldn't be merged - it was just a hack to allow for displaying of raw airspeed on ground for debugging

Comment (kd0aij) [reply ArduPlane/ArduPlane.cpp]: Sorry, I should have mentioned that that commit has to come out before merging

File: ArduPlane/ArduPlane.cpp (hunk 2 above)
Comment (kd0aij) [reply ArduPlane/ArduPlane.cpp]: one Hz was not entertaining enough :) I wanted to see what the noise looked like

**Comment by kd0aij** [comment]
@magicrub  In the second plot above, all the negative dynamic pressure peaks would have shown up as positive airspeed and pushed the average airspeed up significantly

**Comment by kd0aij** [comment]
@tridge Assuming you meant: 
https://github.com/kd0aij/ardupilot/blob/airspeed_fix/libraries/AP_Airspeed/Airspeed_Calibration.cpp#L128

I think it might be better to use max(press, 0) to avoid sqrt(-ve) since that would be closer to the correct (zero) airspeed.  But I'd hope that airspeed during autocal is high enough to stay positive otherwise the SNR would be very low.

**Comment by tridge** [comment]
> @tridge Assuming you meant:

yes, thats the one
Can you fixup the PR ready to merge? I'd like to bring this into 3.8.3

**Comment by kd0aij** [comment]
@tridge On the road for the next few days, and having problems building on my Chromebook, so unable to test that last commit. Should be OK if CI passes, though.

**Comment by kd0aij** [comment]
@tridge Not sure what's wrong with CI; builds OK for me on sitl and px4-v4

**Comment by WickedShell** [comment]
```
[ 548/1401] Compiling libraries/DataFlash/examples/DataFlash_AllTypes/DataFlash_AllTypes.cpp
In file included from ../../libraries/AP_RangeFinder/examples/RFIND_test/RFIND_test.cpp:6:0:
../../libraries/AP_RangeFinder/RangeFinder_Backend.h: In member function ‘MAV_DISTANCE_SENSOR AP_RangeFinder_Backend::get_mav_distance_sensor_type() const’:
../../libraries/AP_RangeFinder/RangeFinder_Backend.h:47:20: error: ‘MAV_DISTANCE_SENSOR_UNKNOWN’ was not declared in this scope
             return MAV_DISTANCE_SENSOR_UNKNOWN;
                    ^
compilation terminated due to -Wfatal-errors.
```

Looks like a MAVLink related issue, I'd try rebasing on master and see if that resolves it.

**Comment by magicrub** [comment]
The px4-v3 build failure shows:
/home/travis/build/ArduPilot/ardupilot/libraries/AP_AccelCal/AP_AccelCal.cpp: In member function 'void AP_AccelCal::update()':
/home/travis/build/ArduPilot/ardupilot/libraries/AP_AccelCal/AP_AccelCal.cpp:166:58: error: 'ACCELCAL_VEHICLE_POS_SUCCESS' was not declared in this scope
                     _gcs->send_accelcal_vehicle_position(ACCELCAL_VEHICLE_POS_SUCCESS);

**Comment by magicrub** [comment]
definitely needs a rebase, its showing submodule changes to mavlink and uavcan.

**Comment by WickedShell** [comment]
@kd0aij this commit https://github.com/ArduPilot/ardupilot/pull/6980/commits/4cf0d2a84631d063f3f8c15c55af74c4d33cdc9a updates the mavlink/uavcan repos which is wrong (and I'm assuming entirely inadvertent). If you remove the changes to those 2 files you should be passing CI again.

**Comment by tridge** [comment]
I've fixed up the autocal and mavlink commit and pushed to master
many thanks for finding this Mark!

**Comment by kd0aij** [comment]
@tridge @WickedShell Thanks, I never noticed that commit was screwed up

**Pull request closed without merging**